garage/misc/wrappers.py

```diff
@@ -0,0 +1,88 @@
+"""Gym env wrappers"""
+import cv2
+import gym
+import numpy as np
+
+from collections import deque
+from gym.spaces import Box
+
+
+class RepeatActionWrapper(gym.core.Wrapper):
+    """Gym env wrapper for repeating action for n frames"""
+
+    def __init__(self, env, frame_to_repeat):
+        super(RepeatActionWrapper, self).__init__(env)
+        self.frame_to_repeat = frame_to_repeat
+
+    def step(self, action):
+        for i in range(self.frame_to_repeat):
+            obs, reward, done, info = self.env.step(action)
+        return obs, reward, done, info
+
+    def reset(self):
+        return self.env.reset()
+
+
+class ResizeAndGrayscaleWrapper(gym.core.Wrapper):
+    """Wrapper for resizing to (w, h) and converting frames to grayscale."""
+
+    def __init__(self, env, w, h, plot=False):
+        super(ResizeAndGrayscaleWrapper, self).__init__(env)
+        self.observation_space = Box(0.0, 1.0, shape=[w, h], dtype=np.float32)
+        self.w = w
+        self.h = h
+        self.plot = plot
+
+    def _observation(self, obs):
+        obs = cv2.cvtColor(obs, cv2.COLOR_RGB2GRAY)
+        obs = cv2.resize(obs, (self.w, self.h), interpolation=cv2.INTER_AREA)
+        obs = obs.astype(np.float32) / 255.0
+        if self.plot:
+            self._display(obs)
+        return obs
+
+    def reset(self):
+        return self._observation(self.env.reset())
+
+    def step(self, action):
+        obs, reward, done, info = self.env.step(action)
+        return self._observation(obs), reward, done, info
+
+    def _display(self, obs):
+        cv2.imshow('image', np.squeeze(np.asarray(obs)))
+        cv2.waitKey(0)
+
+
+class StackFramesWrapper(gym.core.Wrapper):
+    """
+    Gym env wrapper to stack multiple frames.
+    Useful for training feed-forward agents on dynamic games.
+    """
+
+    def __init__(self, env, n_frames_stacked):
+        super(StackFramesWrapper, self).__init__(env)
+        if len(env.observation_space.shape) != 2:
+            raise Exception('Stack frames works with 2D single channel images')
+        self._n_stack_past = n_frames_stacked
+        self._frames = None
+
+        new_obs_space_shape = env.observation_space.shape + (
+            n_frames_stacked, )
+        self.observation_space = Box(
+            0.0, 1.0, shape=new_obs_space_shape, dtype=np.float32)
+
+    def _frames_as_numpy(self):
+        np_frames = np.asarray(self._frames)
```

Comment:
This allocates new memory, but we already have memory allocated for the frame (in `observation`). Can't we concatenate the frames we already have and return them?